### PR TITLE
oc: adding integration tests for 'new-app --binary'

### PR DESCRIPTION
**Pull request by hferentschik**
See also https://trello.com/c/XJSisZuD

### CODE REVIEW

**Comment by hferentschik** [comment]
@bparees WDYT?

File: pkg/oc/lib/newapp/cmd/resolve_test.go [hunk 1]
```diff
@@ -54,6 +54,40 @@ func TestResolveDockerfileAndSource(t *testing.T) {
...
+		t.Fatal(err)
+	}
+
+	if component.Uses == nil {
+		t.Fatal("expected source repository to be created")
+	}
+
+	if component.Uses == nil {
+		t.Fatal("expected source repository to be created")
+	}
```
Comment (bparees) [inline pkg/oc/lib/newapp/cmd/resolve_test.go]: why is this here twice?

File: test/cmd/newapp.sh [hunk 2]
```diff
@@ -5,19 +5,25 @@ trap os::test::junit::reconcile_output EXIT
 # Cleanup cluster resources created by this test
 (
   set +e
-#  oc delete all,templates --all
-  oc delete-project template-substitute
-  oc delete-project prefix-template-substitute
+  oc login -u system:admin
+  oc delete all,templates --all
+  oc delete all,templates --all -n openshift
```
Comment (bparees) [inline test/cmd/newapp.sh]: this is potentially destructive to other tests using this cluster.

File: test/cmd/newapp.sh [hunk 3]
```diff
@@ -612,5 +610,23 @@ os::cmd::expect_success 'oc new-app php-55-centos7:latest https://github.com/ope
 os::cmd::expect_success_and_text 'oc get bc test8 --template={{.spec.strategy.sourceStrategy.from.name}}' 'php-55-centos7:latest'
 os::cmd::expect_success 'oc delete project new-app-syntax'
 
+# new-app docker build strategy with binary input
+os::cmd::expect_success 'oc project ${default_project}'
+os::cmd::expect_success 'oc delete all,templates --all'
+os::cmd::expect_success 'oc create -f examples/image-streams/image-streams-centos7.json'
+os::cmd::try_until_success 'oc get imagestreamtags ruby:2.0' # need to wait until tags are available!?
```
Comment (bparees) [inline test/cmd/newapp.sh]: you should at least be waiting until the tag you care about is available.  2.0 is a bad choice given how old it is.

File: test/cmd/newapp.sh [hunk 4]
```diff
@@ -612,5 +610,23 @@ os::cmd::expect_success 'oc new-app php-55-centos7:latest https://github.com/ope
...
+os::cmd::expect_failure_and_text 'oc new-app --strategy=docker --name my-docker-app' 'none of the arguments provided could be classified as a source code location'
+os::cmd::expect_success_and_text 'oc new-app --strategy=docker --binary --name my-docker-app' 'A binary build was created'
+os::cmd::expect_success_and_text 'oc start-build my-docker-app --from-dir=./test/testdata/testapp' 'build.build.openshift.io/my-docker-app-1 started'
```
Comment (bparees) [inline test/cmd/newapp.sh]: you don't need to run the build.

File: test/cmd/newapp.sh [hunk 5]
```diff
@@ -612,5 +610,23 @@ os::cmd::expect_success 'oc new-app php-55-centos7:latest https://github.com/ope
...
+os::cmd::expect_success 'oc delete all -l app=my-docker-app'
+
+# new-app source build strategy with binary input
+os::cmd::expect_success_and_text 'oc new-app  ruby --binary --name my-imagestream-app' 'A binary build was created'
+os::cmd::expect_success_and_text 'oc start-build my-imagestream-app --from-dir=./test/testdata/testapp' 'build.build.openshift.io/my-imagestream-app-1 started'
```
Comment (bparees) [inline test/cmd/newapp.sh]: don't need to run the build

File: test/cmd/newapp.sh [hunk 6]
```diff
@@ -165,16 +167,15 @@ os::cmd::expect_failure_and_text 'oc new-app -f test/testdata/template-with-name
...
-os::cmd::expect_success 'oc project ${project}'
+os::cmd::expect_success 'oc project ${default_project}'
 os::cmd::expect_success 'oc new-app -f test/testdata/template-with-namespaces.json -p SUBSTITUTED=template-substitute'
 os::cmd::expect_success 'oc delete all -l app=ruby-helloworld-sample'
 
 # ensure non-duplicate invalid label errors show up
-os::cmd::expect_failure_and_text 'oc new-app docker.io/library/wordpress -l qwer1345%$$#=self' 'error: ImageStream.image.openshift.io "wordpress" is invalid'
-os::cmd::expect_failure_and_text 'oc new-app docker.io/library/wordpress -l qwer1345%$$#=self' 'DeploymentConfig.apps.openshift.io "wordpress" is invalid'
+os::cmd::expect_failure_and_text 'oc new-app docker.io/library/wordpress -l qwer1345%$$#=self' 'error: ImageStream "wordpress" is invalid'
+os::cmd::expect_failure_and_text 'oc new-app docker.io/library/wordpress -l qwer1345%$$#=self' 'DeploymentConfig "wordpress" is invalid'
```
Comment (bparees) [inline test/cmd/newapp.sh]: why did you change this? it has nothing to do w/ your changes that i can see.

File: test/cmd/newapp.sh [hunk 7]
```diff
@@ -89,9 +92,8 @@ os::cmd::expect_success 'oc delete all -l app=expose-output'
 os::cmd::expect_success_and_text 'oc new-app mysql --as-test' 'Application is not exposed'
 os::cmd::expect_success 'oc delete all -l app=mysql'
 
-# ensure that oc new-app does not emit a BuildConfigInstantiateFailed event when creating
-# a new application
-os::cmd::expect_success 'oc new-app https://github.com/sclorg/ruby-ex'
+# ensure that oc new-app does not emit a BuildConfigInstantiateFailed event when creating a new application
+os::cmd::expect_success 'oc new-app https://github.com/sclorg/ruby-ex --strategy=source'
```
Comment (bparees) [inline test/cmd/newapp.sh]: why was this changed?

File: pkg/oc/lib/newapp/cmd/resolve_test.go (hunk 1 above)
Comment (hferentschik) [reply pkg/oc/lib/newapp/cmd/resolve_test.go]: My mistake. I'll remove it.

File: test/cmd/newapp.sh (hunk 2 above)
Comment (hferentschik) [reply test/cmd/newapp.sh]: Agreed, I also find it a bit odd to make use of the _openshift_ namespace here. However, this is not something I introduced, but this is already happening in the test see https://github.com/openshift/origin/blob/master/test/cmd/newapp.sh#L21 and https://github.com/openshift/origin/blob/master/test/cmd/newapp.sh#L57. Initially, the test run would fail pretty much immediately since I had templates imported in the _openshift_ namespace. This change makes sure that the right preconditions for this test suite are met. It might actually make more sense to use a dedicated namespace, but I am not sure whether there was a deeper thought behind choosing _openshift_ in the first place. 
FAICT, this does not impact the rest of the integration tests atm.

Should I change this to use another namespace?

File: test/cmd/newapp.sh (hunk 4 above)
Comment (hferentschik) [reply test/cmd/newapp.sh]: OK. My intention was to show no build was automatically started when creating the app. But I can remove it.

On another note, I was tempted to split all start-build commands into its own startbuild.sh test suite. They are really two different commands and this test suite is already so big and takes a long time to run. WDYT?

File: test/cmd/newapp.sh (hunk 5 above)
Comment (hferentschik) [reply test/cmd/newapp.sh]: OK

File: test/cmd/newapp.sh (hunk 3 above)
Comment (bparees) [reply test/cmd/newapp.sh]: it's intended, polling/waiting until the tag is imported is the right thing to do.

File: test/cmd/newapp.sh (hunk 2 above)
Comment (bparees) [reply test/cmd/newapp.sh]: I think the reason for using the openshift namespace is that is where new-app looks for imagestreamtags, in addition to looking in the current namespace.

i think you can leave this.

Comment (hferentschik) [reply test/cmd/newapp.sh]: ok

File: test/cmd/newapp.sh (hunk 3 above)
Comment (hferentschik) [reply test/cmd/newapp.sh]: ok

File: test/cmd/newapp.sh (hunk 6 above)
Comment (hferentschik) [reply test/cmd/newapp.sh]: I tried to use my locally built `oc` binary to run `cluster up`, but I cannot get this to work. I am running into various roadblocks when I do that. For example, I seem to be running into this issue - https://github.com/openshift/origin/issues/20263

I guess I would need some help on how to start a cluster using `oc` build from HEAD (including the use of locally build OpenShift images). I might not have the right workflow for this. 

Running the tests in the following setup:
```
$ oc version
oc v4.0.0-alpha.0+da45ed5-193-dirty
kubernetes v1.11.0+d4cacc0
features: Basic-Auth

Server https://192.168.99.109:8443
openshift v3.10.0+fd501dd-48
kubernetes v1.10.0+b81c8f8
```
I need to change the test assertions. Neither 'image.openshift.io' nor '.apps.openshift.io' appears in the output.

Not sure how to go about this. Are there any CI builds which run these tests? Just to verify that they do pass.

**Comment by hferentschik** [comment]
@bparees I pushed an update for this pull request addressing the points we discussed. The remaining question is https://github.com/openshift/origin/pull/21111/commits/de099c999b7eb19cca9cfa7d3120ab0183eb5ba9#diff-1d1640f9dfba9f6d7b6368be31f1bdabR177. For all I can tell the existing assertion is not correct.

**Comment by bparees** [comment]
> For all I can tell the existing assertion is not correct.

the existing assertion must be correct or the test would be failing, and that test is run against all origin PRs.

i will run the tests against your current PR, and i think we will see it fail.

/ok-to-test

File: test/cmd/newapp.sh (hunk 6 above)
Comment (bparees) [reply test/cmd/newapp.sh]: it behaves as expected with a 3.11 server for me:

```
$ oc version
oc v4.0.0-alpha.0+c2e7c2a-184
kubernetes v1.11.0+d4cacc0
features: Basic-Auth GSSAPI Kerberos SPNEGO

Server https://api.ci.openshift.org:443
openshift v3.11.0-alpha.0+010674d-1220
kubernetes v1.11.0+d4cacc0
```

```
--> Creating resources with label  ...
    error: ImageStream.image.openshift.io "wordpress" is invalid: metadata.labels: Invalid value: "qwer1345%3902#": name part must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]')
    error: DeploymentConfig.apps.openshift.io "wordpress" is invalid: [metadata.labels: Invalid value: "qwer1345%3902#": name part must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]'), spec.template.labels: Invalid value: "qwer1345%3902#": name part must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]')]
    error: Service "wordpress" is invalid: [metadata.labels: Invalid value: "qwer1345%3902#": name part must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]'), spec.selector: Invalid value: "qwer1345%3902#": name part must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]')]
```

Comment (bparees) [reply test/cmd/newapp.sh]: and yeah, our CI jobs that run against PRs run these tests, so your PR won't merge if they are not correct.

File: test/cmd/newapp.sh [hunk 8]
```diff
@@ -612,5 +610,21 @@ os::cmd::expect_success 'oc new-app php-55-centos7:latest https://github.com/ope
...
+os::cmd::expect_success 'oc project ${default_project}'
+os::cmd::expect_success 'oc delete all,templates --all'
+os::cmd::expect_success 'oc create -f examples/image-streams/image-streams-centos7.json'
+os::cmd::try_until_success 'oc get imagestreamtags ruby:latest' # need to wait until tags are available!?
+os::cmd::expect_failure_and_text 'oc new-app --strategy=docker --name my-docker-app' 'none of the arguments provided could be classified as a source code location'
+os::cmd::expect_success_and_text 'oc new-app --strategy=docker --binary --name my-docker-app' 'A binary build was created'
+os::cmd::expect_success 'oc delete all -l app=my-docker-app'
+
+# new-app source build strategy with binary input
+os::cmd::expect_success_and_text 'oc new-app  ruby --binary --name my-imagestream-app' 'A binary build was created'
```
Comment (bparees) [inline test/cmd/newapp.sh]: you should confirm this created a source strategy build

File: test/cmd/newapp.sh [hunk 9]
```diff
@@ -612,5 +610,21 @@ os::cmd::expect_success 'oc new-app php-55-centos7:latest https://github.com/ope
 os::cmd::expect_success_and_text 'oc get bc test8 --template={{.spec.strategy.sourceStrategy.from.name}}' 'php-55-centos7:latest'
 os::cmd::expect_success 'oc delete project new-app-syntax'
 
+# new-app docker build strategy with binary input
+os::cmd::expect_success 'oc project ${default_project}'
+os::cmd::expect_success 'oc delete all,templates --all'
+os::cmd::expect_success 'oc create -f examples/image-streams/image-streams-centos7.json'
+os::cmd::try_until_success 'oc get imagestreamtags ruby:latest' # need to wait until tags are available!?
+os::cmd::expect_failure_and_text 'oc new-app --strategy=docker --name my-docker-app' 'none of the arguments provided could be classified as a source code location'
+os::cmd::expect_success_and_text 'oc new-app --strategy=docker --binary --name my-docker-app' 'A binary build was created'
```
Comment (bparees) [inline test/cmd/newapp.sh]: you should confirm this created a docker strategy build

Comment (hferentschik) [reply test/cmd/newapp.sh]: ok

File: test/cmd/newapp.sh (hunk 8 above)
Comment (hferentschik) [reply test/cmd/newapp.sh]: ok

**Comment by hferentschik** [comment]
@bparees now the cmd test pass. Not quite sure what the issue is with extended_clusterup. Looking at the output it seems that the cluster does not even come up. Seems unrelated.

**Comment by bparees** [comment]
@hferentschik i'm not sure why that one is failing either.  Squash your commits and i'll put an lgtm on it and hopefully the issue resolves itself.

**Comment by hferentschik** [comment]
>  i'm not sure why that one is failing either. Squash your commits and i'll put an lgtm on it and hopefully the issue resolves itself.

@bparees rebased and squashed

File: test/cmd/newapp.sh (hunk 9 above)
Comment (bparees) [reply test/cmd/newapp.sh]: this isn't checking the strategy (s2i vs docker)

File: test/cmd/newapp.sh (hunk 8 above)
Comment (bparees) [reply test/cmd/newapp.sh]: this isn't checking the strategy (s2i vs docker)

**Comment by bparees** [comment]
thanks @hferentschik
